### ISSUE TITLE
Create a Pester function ShouldBeNonTerminatingErrorId to complement ShouldBeErrorId

### ISSUE THREAD

**Issue by dantraMSFT** [PowerShell/PowerShell#4825]
The purpose of this function is to encapsulate verifying a non-terminating error just as ShouldBeErrorId verifies terminating errors.

The reason for the function is to avoid tests using -ErrorAction Stop to leverage ShouldBeErrorId.  While this works, it does not verify that the error is, in fact non-terminating and an intentional or unintentional in the underlying API would be unnoticed.

it also clarifies that the test is expecting a non-terminating error.

**Comment by iSazonov** [comment]
If we add a property as requested in #4781 we can easy check that an error is terminating error:
```powershell
$exc = { Select-XML -ErrorAction Stop } | ShouldBeErrorId "..."
if ($exc.IsTerminating) { ... }
```
With this approach, we can continue to use one pattern but be more accurate in tests.
I think that using a new function and a new pattern will complicate our tests. While the new property `IsTerminating` will give us more opportunities.

**Comment by SteveL-MSFT** [comment]
Perhaps:

```powershell
{ select-xml whatever } | ShouldBeErrorId "..." | ShouldBeTerminatingError
```

Agree it should be built upon #4781 which isn't planned for 6.0.0.

**Comment by dantraMSFT** [comment]
I was thinking along the lines of the following:

{ select-xml whatever } | ShouldBeErrorId "..." -Nonterminating

**Comment by iSazonov** [comment]
I guess Pester unlike parameters `-Nonterminating` for `Should`.
(Also next version of Pester have a implementation of `ShouldBeErrorId`. )

**Comment by dantraMSFT** [comment]
@iSazonov: I have to disagree with the  IsTerminating property for solving this problem (test verifying non-terminating errors).  If the error is non-terminating, the error stream contains the error; otherwise, the error is thrown. For both cases, the property is moot from a test perspective.  Of course, a function for testing terminating or non-terminating errors should be prepared for detecting the incorrect error pattern. 

Currently, ShouldBeErrorId does not do that.

**Comment by iSazonov** [comment]
@dantraMSFT Good point! I agree that  IsTerminating property don't resolve the problem and we should think about new pattern.

So what is a conclusion? - use current pattern until we start to resolve #4781?

**Comment by SteveL-MSFT** [comment]
For now, I think we should continue to use the pattern until we resolve #4781 so that we can easily find these occurrences in the future and fix them in bulk

**Comment by dantraMSFT** [comment]
My opinion is it should be independent; the testing requirement can be solved without a product change.  I've mocked up -Nonterminating and extended ShouldBeErrorId to handle both as well as fixing a subtle problem with the try/catch.  If people don't like the -Nonterminating switch, a new function could be created.
Both should handle the case where the test expects terminating/non-terminating errors but the product does the opposite; though.

**Comment by SteveL-MSFT** [comment]
cc @JamesWTruher your thoughts on the switch?

**Comment by bergmeister** [comment]
I find the already existing `ShouldBeErrorId` assertion function odd because it is not part of vanilla Pester. If any additional assertion function is generally useful (and I think at least `ShouldBeErrorId`  is), then you should rather open a PR and add it to Pester itself. The assertion function not being part of vanilla Pester makes it harder to run tests locally by just running `Invoke-Pester` against the test file. You must also think of the use case that someone only wants to make a small contribution and repro a failing test in CI locally without having to do too much setup work and look for documentation.

**Comment by markekraus** [comment]
@bergmeister all tests in this project depend on using `Start-PSPester` to ensure shared code, binaries, pathes, and modules are present. I don't think there is currently any expectation to run project tests with `Invoke-Pester` directly.  You can call a specific test or tag directly with `Start-PSPester`. 

I use this often when testing the web cmdlet changes on Windows:

```PowerShell
Start-PSPester -powershell .\src\powershell-win-core\bin\Release\netcoreapp2.0\win7-x64\publish\pwsh.exe -binDir .\src\powershell-win-core\bin\Release\netcoreapp2.0\win7-x64\publish\  -Path .\test\powershell\Modules\Microsoft.PowerShell.Utility\WebCmdlets.Tests.ps1 -Terse
```

**Comment by markekraus** [comment]
On another note, I think that both functionalities should be a part of Pester and not something we and 100's of other projects have to re-invent the wheel on.

**Comment by SteveL-MSFT** [comment]
@markekraus by default, `Start-PSPester` should automatically figure out the location of your last build, so you shouldn't need to use the `-powershell` switch or the `-bindir` switch.

I agree that the validation of errors is generic and we should have this discussion in Pester repo and (hopefully) eventually remove our custom function.

**Comment by markekraus** [comment]
It should... but doesn't all of the time. I'm not sure why. Rather than dig into it I just keep that handy in my history.

**Comment by iSazonov** [comment]
What is Perter repo Issue number?

**Comment by SteveL-MSFT** [comment]
@dantraMSFT can you open a new issue in Pester repo specifically to handle terminating and non-terminating errors with `Should -Throw`?

**Comment by bergmeister** [comment]
@SteveL-MSFT  In pester 4.2 (see blog [here](http://jakubjares.com/2017/12/19/using-should-throw/)) it is now possible to assert against the errorId (example taken from blog):
````powershell
    It 'throws argument exception when given $null' {
        { Get-Computer `$null } |
            Should -Throw `
                -ExceptionType ([ArgumentException]) `
                -ErrorId 'SpecificErrorId'
    }
````
Also, I cannot see an issue being raised about terminating vs non-terminating errors in Pester (but I do approve that this would be very helpful).

**Comment by markekraus** [comment]
@bergmeister yea. We need to update our guidance and then begin the process of refactoring. When the the project switched PSGallery Pester we did the minimal just to get the existing tests working under the new pester, but, we should also refactor existing tests to fit the new ways of doing things. (I.e. using `-Be` instead of `Be` and some of the newer options like `-HaveCount`, `-BeTrue`, and `-BeFalse` and `-Throw -ErrorId`)

**Comment by iSazonov** [comment]
Is it style changes or we get any benefits (performance, remove deprecated Pester patterns)?

**Comment by markekraus** [comment]
Refactoring tests for consistency, enforcement of best practices, and the ability to publish and enforce guidance are my concern. I'm tired of the excuse that we have to shove new ways of doing things into old and rusted code patterns just because refactoring does "provide any benefits". I can understand not refactoring the code base for those reasons, but tests? We don't need to justify doing things correctly, IMO.

**Comment by bergmeister** [comment]
the benefit to me is to move away rather than into a home-grown test framework so that the majority of tests can be executed using vanilla Pester. I have already upgraded all `PSScriptAnalyzer` operators to v4 syntax and to the new operators in 4.2 and can say that they work fine except for `HasCount`, which I raised a bug.

**Comment by iSazonov** [comment]
I think we need a new tracking Issue with the rationales to get confirmation from MSFT team.

**Comment by bergmeister** [comment]
We probably need a new issue for some of the stuff that we discussed but this issue should also be re-opened because it seemed to be closed by accident.

**Comment by SteveL-MSFT** [comment]
Now, I remember.  I closed this as we should discuss validating terminating vs non-terminating in the Pester repo rather than having another custom function.  I agree with the other statements that we should move away from custom functions as much as possible.  Less code for us to maintain and also community contributors can use existing patterns from Pester rather than learning something unique to this project.  We should convert our existing tests to use the new Pester support for error validation.  Their latest release supports `-PassThru` so I think everything is needed to remove `ShouldBeErrorId`.  Closing again.
